### PR TITLE
storage: add strawman re-implementation of core functionality

### PR DESCRIPTION
**Pull request by henrytill**
Ref:
https://rchain.atlassian.net/wiki/spaces/CORE/pages/192905321/Storage.01+Specification+WIP

As I started sketching out the core storage API in code, it quickly turned into an actual implementation.

I am submitting this code as a strawman proposal for a new beginning to the storage project.  Consider it a fresh start to iterate on.

Feedback welcome!

### CODE REVIEW

**Comment by kirkwood** [comment]
The protocol buffer business is there by way of example (and to give you some fodder for running the thing), right?

**Comment by henrytill** [comment]
> The protocol buffer business is there by way of example (and to give you some fodder for running the thing), right?

Correct - the message itself is just a stand-in for the real thing. 

@pyrocto did say that he wanted to use protobuf as the serialization format.

**Comment by kirkwood** [comment]
Oh, well, that simplifies my life.

File: storage/src/main/scala/coop/rchain/storage/Storage.scala [hunk 1]
```diff
@@ -1,266 +1,76 @@
...
+    extends IStorage
+    with AutoCloseable {
+
+  def put[A](key: Key, value: A)(implicit s: Serialize[A]): Either[Error, Unit] =
+    Either
+      .catchNonFatal {
+        val encoded = s.encode(value)
+        val keyBuff = ByteBuffer.allocateDirect(env.getMaxKeySize)
+        val valBuff = ByteBuffer.allocateDirect(encoded.length)
+        ignore { keyBuff.put(key.bytes).flip() }
```
Comment (pyrocto) [inline storage/src/main/scala/coop/rchain/storage/Storage.scala]: Why use `ignore`?

Comment (henrytill) [reply storage/src/main/scala/coop/rchain/storage/Storage.scala]: I want to be explicit about the fact that this (impure) expression returns a value (in this case for the sake of method-chaining) that I am discarding, and am instead only using for its side-effects.

We might imagine doing something like this:
```scala
val _ = keyBuff.put(key.bytes).flip()
val _ = valBuff.put(encoded).flip()
```

However, `_` is a normal identifier in Scala, and we can't rebind variable names.  So instead we have to do something like:
```scala
val _1 = keyBuff.put(key.bytes).flip()
val _2 = valBuff.put(encoded).flip()
```
...which is kind of awful, and will also trigger compiler and IDE warnings about unused variables. 

So instead, I borrowed the `ignore` function from the OCaml standard library.

I'd also like to start using the [wartremover](http://www.wartremover.org/) code linting tool on our code, which will complain about cases where we are discarding return values without explicitly accounting for doing so.

**Pull request merged**